DOC Add tutorial for adding a new package (#1829)
pyodide/pyodide: docs/development/new-packages.md

# Creating a Pyodide package

## Quickstart

If you wish to use a package in Pyodide that is not already included, first you

> need to determine if it is necessary to package if for Pyodide. Ideally, you

Reviewer: ```suggestion
need to determine whether it is necessary to package it for Pyodide. Ideally, you
```

If the wheel is not on PyPi, but nevertheless you believe there is nothing
preventing it (it is a Python package without C extensions):

- you can create the wheel yourself by running,
  ```
  python -m pip install build
  python -m build
  ```
  from within the package folder where the `setup.py`
  are located. See the [Python packaging
  guide](https://packaging.python.org/tutorials/packaging-projects/#generating-distribution-archives)
  for more details.
  Then upload the wheel file somewhere (not to PyPi) and install it with
  micropip via its URL.
- you can also open an issue in the package repository asking the
  authors to upload the wheel.

If however the package has C extensions or its code requires patching, then
continue to the next steps.

```{note}
To determine if a package has C extensions, check if its `setup.py` contains
any compilation commands.
```

### 2. Creating the meta.yaml

Once you determined that you need to create a new package for Pyodide, the
easiest place to start is with the {ref}`mkpkg tool <pyodide-mkpkg>`. If your
package is on PyPI, run:

`pyodide-build mkpkg <package-name>`

This will generate a `meta.yaml` under `package/<package-name>/` (see

> `ref`{meta-yaml-spec}) that should work out of the box for many simple Python

Reviewer: ```suggestion
{ref}`meta-yaml-spec`) that should work out of the box for many simple Python
```

will need to specify those yourself.

You can also use the `meta.yaml` of other Pyodide packages in the `packages/`
folder as a starting point.

```{note}
To reliably determine build and runtime dependencies, including for non Python
libraries, it is often useful to verify if the package was already built on
[conda-forge](https://conda-forge.org/) and open the corresponding `meta.yaml`
file. This can be done either by checking if the URL
`https://github.com/conda-forge/<package-name>-feedstock/blob/master/recipe/meta.yaml`
exists, or by searching the [conda-forge Github
org](https://github.com/conda-forge/) for the package name.

The `meta.yaml` in Pyodide was inspired by the one in conda, however it is
not strictly compatible.
```

### 3. Building the package and investigating issues

Once the `meta.yaml` is ready, we build the package with,

```
PYODIDE_PACKAGES="<package-name>" make
```

and see if there are any errors. The detailed build log can be found under
`packages/<package-name>/<package-name>.log`.

If there are errors you might need to,

- patch the package by adding `.patch` files to `packages/<package-name>/patches`

> - update the `meta.yaml`

Reviewer: Maybe more specifically 
```suggestion
- add the patch files to the `source/patches` field in the `meta.yaml` file
```

then re-start the build.

> In general, it is useful to see how other similar packages are built in Pyodide.

Reviewer: ```suggestion
In general, it is recommended to look into how other similar packages are built in Pyodide.
```
